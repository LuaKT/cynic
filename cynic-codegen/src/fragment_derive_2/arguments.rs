use proc_macro2::TokenStream;
use syn::{
    ext::IdentExt,
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    Expr, Ident, Result, Token,
};

pub fn arguments_from_field_attrs(attrs: &[syn::Attribute]) -> Result<Vec<FieldArgument>> {
    for attr in attrs {
        if attr.path.is_ident("arguments") {
            let parsed: CynicArguments = attr.parse_args()?;
            return Ok(parsed.arguments.into_iter().collect());
        }
    }
    Ok(vec![])
}

/// Implements syn::Parse to parse out arguments from the arguments
/// attribute.
#[derive(Debug)]
struct CynicArguments {
    // TODO: technically we want some kind of MaybePunctuated.
    // Worth looking into that later
    arguments: Punctuated<FieldArgument, Token![,]>,
}

impl Parse for CynicArguments {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(CynicArguments {
            arguments: Punctuated::parse_terminated(input)?,
        })
    }
}

#[derive(Debug, Clone)]
pub struct FieldArgument {
    pub argument_name: Ident,
    pub value: ArgumentLiteral,
}

impl Parse for FieldArgument {
    fn parse(input: ParseStream) -> Result<Self> {
        let argument_name = input.parse()?;
        input.parse::<Token![:]>()?;
        let value = input.parse()?;

        Ok(FieldArgument {
            argument_name,
            value,
        })
    }
}

// TODO: Think about spans here...
#[derive(Debug, Clone)]
pub enum ArgumentLiteral {
    // True & false are idents aren't they, fuck
    True,
    False,
    Literal(proc_macro2::Literal),
    Object(Punctuated<FieldArgument, Token![,]>),
    List(Punctuated<ArgumentLiteral, Token![,]>),
    Variable(proc_macro2::Ident),
    Null,
}

impl Parse for ArgumentLiteral {
    fn parse(input: ParseStream) -> Result<Self> {
        let lookahead = input.lookahead1();
        if lookahead.peek(syn::token::Brace) {
            let content;
            syn::braced!(content in input);
            let obj = Punctuated::<FieldArgument, Token![,]>::parse_terminated(&content)?;

            // TODO: ideally return something other than punctuated, but need to retain spans.
            Ok(ArgumentLiteral::Object(obj))
        } else if lookahead.peek(syn::token::Bracket) {
            let content;
            syn::bracketed!(content in input);
            let list = Punctuated::<ArgumentLiteral, Token![,]>::parse_terminated(&content)?;

            Ok(ArgumentLiteral::List(list))
        } else if lookahead.peek(Token![$]) {
            input.parse::<Token![$]>()?;

            Ok(ArgumentLiteral::Variable(input.parse()?))
        } else if lookahead.peek(Ident::peek_any) {
            input.parse::<Token![$]>()?;
            let ident = input.call(Ident::parse_any)?;

            // TODO: Could be true, false, null or an error?
            todo!()
        } else if lookahead.peek(syn::Lit) {
            Ok(ArgumentLiteral::Literal(input.parse()?))
        } else {
            Err(lookahead.error())
        }
    }
}


/*
impl quote::ToTokens for FieldArgument {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        use quote::{quote, TokenStreamExt};

        let argument_name = &self.argument_name;
        let expr = &self.expr;

        tokens.append_all(quote! {
            #argument_name: #expr
        });
    }
} */

#[cfg(test)]
mod test {
    use assert_matches::assert_matches;
    use syn::parse_quote;

    use super::*;

    #[test]
    fn test_parsing_string_literal() {
        let parsed: CynicArguments = parse_quote! { x: "abcd" };

        let arguments = parsed.arguments.iter().collect::<Vec<_>>();

        assert_eq!(arguments.len(), 1);
        assert_eq!(arguments[0].argument_name.to_string(), "x".to_string());
        assert_matches!(arguments[0].value, ArgumentLiteral::Literal(_));
    }

    #[test]
    fn test_parsing_multiple_arg_expressions() {
        let parsed: CynicArguments = parse_quote! { x: 1, y: $variable };

        let arguments = parsed.arguments.iter().collect::<Vec<_>>();

        assert_eq!(arguments.len(), 2);
        assert_eq!(arguments[0].argument_name.to_string(), "x".to_string());
        assert_matches!(arguments[0].value, ArgumentLiteral::Literal(_));

        assert_eq!(arguments[1].argument_name.to_string(), "y".to_string());
        assert_matches!(&arguments[1].value, ArgumentLiteral::Variable(name) => {
            assert_eq!(name.to_string(), "variable");
        });
    }

    #[test]
    fn test_parsing_list_and_object() {
        let parsed: CynicArguments = parse_quote! { x: {fieldOne: ["hello"], fieldTwo: "hello"}};
        let arguments = parsed.arguments.iter().collect::<Vec<_>>();

        assert_eq!(arguments.len(), 1);
        assert_eq!(arguments[0].argument_name.to_string(), "x".to_string());
        assert_matches!(&arguments[0].value, ArgumentLiteral::Object(fields) => {
            let fields = fields.iter().collect::<Vec<_>>();
            assert_eq!(fields.len(), 2);

            assert_eq!(fields[0].argument_name.to_string(), "fieldOne");
            assert_matches!(&fields[0].value, ArgumentLiteral::List(vals) => {
                let vals = vals.iter().collect::<Vec<_>>();
                assert_eq!(vals.len(), 1);

                assert_matches!(vals[0], ArgumentLiteral::Literal(_));
            });

            assert_eq!(fields[1].argument_name.to_string(), "fieldTwo");
            assert_matches!(fields[1].value, ArgumentLiteral::Literal(_));
        });
    }
}
