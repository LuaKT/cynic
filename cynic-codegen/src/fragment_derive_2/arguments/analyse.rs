use super::parsing;

pub struct Output;

pub fn analyze(literals: Vec<parsing::FieldArgument>, schema: ()) -> Output {
    // Somehow implement this.
    // Though it's not that simple, as I really need more info than the schema parsing in this module
    // currently gives me.

    todo!("analyze not implemented")
}
