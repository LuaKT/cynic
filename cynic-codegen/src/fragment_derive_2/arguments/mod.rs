mod analyse;
mod parsing;

pub use parsing::{arguments_from_field_attrs, FieldArgument};
