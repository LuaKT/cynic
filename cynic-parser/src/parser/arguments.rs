use super::{named_type, selection_set, Parser};
use crate::{lexer::Token, syntax::SyntaxKind::*};

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Context {
    Constant,
    NonConstant,
}

pub(super) fn arguments(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::OpenParen));
    parser.builder.start_node(ARGUMENTS.into());
    parser.bump();
    // TODO: almost wonder if bump should skip ws automatically...
    parser.skip_ws();

    loop {
        match parser.current() {
            None => {
                parser.error("expected arguments");
                break;
            }
            Some(Token::Name) => {
                argument(parser, context);
            }
            Some(Token::CloseParen) => {
                parser.bump();
                parser.skip_ws();
                break;
            }
            _ => {
                // TODO: Is this good?  no idea
                parser.error("expected a name or )");
                break;
            }
        }
    }

    parser.builder.finish_node();
}

fn argument(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(ARGUMENT.into());
    parser.bump();
    parser.skip_ws();

    match parser.current() {
        Some(Token::Colon) => {
            parser.bump();
        }
        _ => {
            parser.error("expected colon");
        }
    }

    value(parser, context);
    parser.skip_ws();
    parser.builder.finish_node();
}

pub(super) fn value(parser: &mut Parser, context: Context) {
    // TODO: Insert a value node here
    parser.builder.start_node(VALUE.into());
    parser.skip_ws();
    match parser.current() {
        Some(Token::Dollar) => {
            if context == Context::Constant {
                parser.error("Found variable in constant context");
                // Continue parsing anyway though...
            }
            super::variable(parser)
        }
        Some(Token::NegativeSign) | Some(Token::Number) => number(parser),
        Some(Token::Quote) | Some(Token::BlockQuote) => string(parser),
        Some(Token::Name) => {
            match parser.current_str().unwrap() {
                "true" | "false" => boolean(parser),
                "null" => null(parser),
                _ => enum_value(parser),
            }
            // Could be boolean, null, or an enum
        }
        Some(Token::OpenCurly) => object(parser, context),
        Some(Token::OpenSquare) => list(parser, context),
        _ => parser.error("expected a value"),
    }
    parser.skip_ws();
    parser.builder.finish_node();
}

fn number(parser: &mut Parser) {
    assert!(matches!(
        parser.current(),
        Some(Token::Number) | Some(Token::NegativeSign)
    ));
    let checkpoint = parser.builder.checkpoint();

    if let Some(Token::NegativeSign) = parser.current() {
        parser.bump();
    }

    match parser.current() {
        Some(Token::Number) => {
            parser.bump();
        }
        _ => parser.error("expected number"),
    }

    if let Some(Token::ExponentPart) | Some(Token::FractionalPart) = parser.current() {
        parser.builder.start_node_at(checkpoint, FLOAT_VALUE.into());
        if let Some(Token::FractionalPart) = parser.current() {
            parser.bump();
        }
        if let Some(Token::ExponentPart) = parser.current() {
            parser.bump();
        }
    } else {
        parser
            .builder
            .start_node_at(checkpoint, INTEGER_VALUE.into());
    }

    parser.builder.finish_node();
}

fn string(parser: &mut Parser) {
    assert!(matches!(
        parser.current(),
        Some(Token::Quote) | Some(Token::BlockQuote)
    ));
    // TODO: probably want to be able to differentiate between normal & block strings.
    parser.builder.start_node(STRING_VALUE.into());

    if parser.current() == Some(Token::Quote) {
        parser.bump();
        loop {
            match parser.current() {
                None => {
                    parser.error("unexpected end of input while parsing string");
                    break;
                }
                Some(Token::Quote) => {
                    parser.bump();
                    break;
                }
                Some(Token::LineTerminator) => {
                    parser.error("missing \"");
                    break;
                }
                _ => {
                    parser.bump_as(STRING_CONTENTS);
                }
            }
        }
    } else {
        parser.bump();
        loop {
            // TODO: I'm pretty sure this doesn't treat escape characters
            // correctly (there's also some mention of escaped block quotes that
            // I don't quite understand - figure this out later.)
            match parser.current() {
                None => {
                    parser.error("unexpected end of input while parsing string");
                    break;
                }
                Some(Token::BlockQuote) => {
                    parser.bump();
                    break;
                }
                _ => {
                    parser.bump_as(STRING_CONTENTS);
                }
            }
        }
    }

    parser.builder.finish_node();
}

fn object(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::OpenCurly));
    parser.builder.start_node(OBJECT_VALUE.into());
    parser.bump();
    parser.skip_ws();

    loop {
        match parser.current() {
            None => {
                parser.error("missing }");
            }
            Some(Token::CloseCurly) => {
                parser.bump();
                break;
            }
            Some(Token::Name) => {
                object_field(parser, context);
            }
            _ => {
                parser.error("expected name");
                break;
            }
        }
    }

    parser.builder.finish_node();
}

fn object_field(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(OBJECT_FIELD.into());
    parser.bump();
    parser.skip_ws();

    match parser.current() {
        Some(Token::Colon) => {
            parser.bump();
        }
        _ => parser.error("expected colon"),
    }
    value(parser, context);

    parser.builder.finish_node();
}

fn list(parser: &mut Parser, context: Context) {
    assert_eq!(parser.current(), Some(Token::OpenSquare));
    parser.builder.start_node(LIST_VALUE.into());
    parser.bump();
    parser.skip_ws();
    loop {
        match parser.current() {
            None => {
                parser.error("missing ]");
                break;
            }
            Some(Token::CloseSquare) => {
                parser.bump();
                break;
            }
            _ => {}
        }
        value(parser, context);
        parser.skip_ws();
    }

    parser.builder.finish_node();
}

fn boolean(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));

    parser.builder.start_node(BOOL_VALUE.into());

    match parser.current_str() {
        Some("true") => parser.bump_as(TRUE),
        Some("false") => parser.bump_as(FALSE),
        _ => panic!("boolean pre-condition not met"),
    }

    parser.builder.finish_node();
}

fn null(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    assert_eq!(parser.current_str(), Some("null"));

    parser.bump_as(NULL);
}

fn enum_value(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));

    parser.builder.start_node(ENUM_VALUE.into());
    parser.bump();
    parser.builder.finish_node();
}
