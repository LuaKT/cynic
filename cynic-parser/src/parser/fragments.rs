use super::{arguments, directives::maybe_directives, named_type, selection_set, Parser};
use crate::{lexer::Token, syntax::SyntaxKind::*};

pub(super) fn fragment(parser: &mut Parser) {
    fragment_name(parser);

    type_condition(parser);

    maybe_directives(parser, arguments::Context::NonConstant);

    parser.skip_ws();
    match parser.current() {
        Some(Token::OpenCurly) => selection_set(parser),
        _ => parser.error("expected selection set"),
    }
}

pub(super) fn fragment_spread(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Spread));
    parser.builder.start_node(FRAGMENT_SPREAD.into());
    parser.bump();
    parser.skip_ws();
    if let Some(Token::Name) = parser.current() {
        fragment_name(parser);
    }

    maybe_directives(parser, arguments::Context::NonConstant);

    parser.skip_ws();
    if parser.current() == Some(Token::OpenCurly) {
        super::selection_set(parser);
    }

    parser.builder.finish_node();
}

fn fragment_name(parser: &mut Parser) {
    parser.skip_ws();
    match parser.current() {
        None => parser.error("expected fragment name"),
        Some(Token::Name) if parser.current_str() != Some("on") => {
            parser.builder.start_node(FRAGMENT_NAME.into());
            parser.bump();
            parser.builder.finish_node();
        }
        _ => {
            parser.error("expected fragment name");
        }
    }
}

pub(super) fn inline_fragment(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Spread));
    parser.builder.start_node(INLINE_FRAGMENT.into());
    parser.bump();
    parser.skip_ws();
    if let Some((Token::Name, "on")) = parser.current_pair() {
        type_condition(parser);
    }

    maybe_directives(parser, arguments::Context::NonConstant);

    parser.skip_ws();
    match parser.current() {
        Some(Token::OpenCurly) => selection_set(parser),
        _ => parser.error("expected selection set"),
    }

    parser.builder.finish_node();
}

fn type_condition(parser: &mut Parser) {
    parser.skip_ws();
    match parser.current_pair() {
        Some((Token::Name, "on")) => {
            parser.builder.start_node(TYPE_CONDITION.into());
            parser.bump_as(ON);
            parser.skip_ws();
            if let Some(Token::Name) = parser.current() {
                named_type(parser);
            } else {
                parser.error("expected named type");
            }
            parser.builder.finish_node();
        }
        _ => {
            parser.error("expected a type condition");
        }
    }
}
