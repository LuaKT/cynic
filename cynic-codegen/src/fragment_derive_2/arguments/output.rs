use proc_macro2::Span;
use quote::{quote, ToTokens, TokenStreamExt};
use syn::Token;

use super::analyse::{AnalysedArguments, ArgumentValue};

pub struct Output<'a> {
    pub(super) analysed: AnalysedArguments<'a>,
    pub(super) schema_module: syn::Path,
}

impl ToTokens for Output<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        if self.analysed.arguments.is_empty() {
            return;
        }

        let argument_module = self
            .analysed
            .schema_field
            .argument_module()
            .to_path(&self.schema_module);

        for arg in &self.analysed.arguments {
            let arg_marker = proc_macro2::Ident::from(arg.schema_field.marker_ident());
            let value = &arg.value;

            tokens.append_all(quote! {
                field_builder.argument::<#argument_module::#arg_marker>()
                #value;
            });
        }
    }
}

impl ToTokens for ArgumentValue<'_> {
    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
        match self {
            ArgumentValue::Object(_) => todo!("obj"),
            ArgumentValue::List(_) => todo!("list"),
            ArgumentValue::Literal(lit) => tokens.append_all(quote! {
                .literal(#lit)
            }),
            ArgumentValue::Bool(b) => {
                let lit = syn::LitBool::new(*b, Span::call_site());
                tokens.append_all(quote! {
                    .literal(#lit)
                });
            }
            ArgumentValue::Variable(_) => todo!("variable"),
            ArgumentValue::Some(_) => todo!("some"),
            ArgumentValue::Null => todo!("null"),
        }
    }
}
