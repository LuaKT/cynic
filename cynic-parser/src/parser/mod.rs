use logos::{Logos, Span};
use rowan::{GreenNode, GreenNodeBuilder};

use crate::{
    lexer::Token,
    syntax::{
        SyntaxKind::{self, *},
        SyntaxNode,
    },
};

mod arguments;
mod fragments;

use arguments::{arguments, constant_value};
use fragments::{fragment, fragment_spread, inline_fragment};

/// The parse results are stored as a "green tree".
/// We'll discuss working with the results later
pub struct Parse {
    green_node: GreenNode,
    #[allow(unused)]
    errors: Vec<(String, Option<Span>)>,
}

pub fn parse(text: &str) -> Parse {
    let mut lexer = Token::lexer(text);
    let mut tokens = Vec::new();
    while let Some(token) = lexer.next() {
        tokens.push((token, lexer.slice(), lexer.span()));
    }
    tokens.reverse();

    Parser {
        tokens,
        builder: GreenNodeBuilder::new(),
        errors: Vec::new(),
    }
    .parse()
}

impl Parse {
    fn syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green_node.clone())
    }
}

struct Parser<'source> {
    tokens: Vec<(Token, &'source str, Span)>,
    builder: GreenNodeBuilder<'static>,

    // TODO: Do I want spanned errors?
    errors: Vec<(String, Option<Span>)>,
}

impl<'source> Parser<'source> {
    fn parse(mut self) -> Parse {
        self.builder.start_node(ROOT.into());
        let mut current_error = false;
        loop {
            // Ok, so parsing out definitions here.
            match executable_def(&mut self) {
                Res::Eof => {
                    break;
                }
                Res::ExpectedExecutableDef => {
                    self.builder.start_node(ERROR.into());
                    if !current_error {
                        self.error("expected executable definition");
                        current_error = true;
                    }
                    self.bump();
                    // TODO: better nodes somehow
                    self.builder.finish_node();
                }
                Res::Ok => {
                    current_error = false;
                }
            }
        }
        self.skip_ws();
        self.builder.finish_node();

        Parse {
            green_node: self.builder.finish(),
            errors: self.errors,
        }
    }

    fn current(&self) -> Option<Token> {
        self.tokens.last().map(|(t, _, _)| *t)
    }

    fn peek_next_non_ws(&self) -> Option<Token> {
        let mut rev_iter = self.tokens.iter().rev();
        rev_iter.next(); // Skip the first one.
        rev_iter
            .skip_while(|(t, _, _)| *t == Token::Whitespace || *t == Token::LineTerminator)
            .next()
            .map(|(t, _, _)| *t)
    }

    fn peek_next_str_non_ws(&self) -> Option<&'source str> {
        let mut rev_iter = self.tokens.iter().rev();
        rev_iter.next(); // Skip the first one.
        rev_iter
            .skip_while(|(t, _, _)| *t == Token::Whitespace || *t == Token::LineTerminator)
            .next()
            .map(|(_, s, _)| *s)
    }

    fn current_str(&self) -> Option<&'source str> {
        self.tokens.last().map(|(_, s, _)| *s)
    }

    fn current_pair(&self) -> Option<(Token, &'source str)> {
        self.tokens.last().map(|(t, s, _)| (*t, *s))
    }

    fn current_span(&self) -> Option<Span> {
        self.tokens.last().map(|(_, _, span)| span.clone())
    }

    fn error(&mut self, err: impl Into<String>) {
        self.errors.push((err.into(), self.current_span()));
    }

    fn bump_as(&mut self, kind: SyntaxKind) {
        let (_, text, _) = self.tokens.pop().unwrap();
        self.builder.token(kind.into(), text)
    }

    fn bump(&mut self) {
        let (kind, text, _) = self.tokens.pop().unwrap();
        self.builder.token(SyntaxKind::from(kind).into(), text);
    }

    fn skip_ws(&mut self) {
        // TODO: This probably needs to know how to deal with comments as well
        while self.current() == Some(Token::Whitespace)
            || self.current() == Some(Token::LineTerminator)
            || self.current() == Some(Token::Comma)
        {
            self.bump();
        }
    }
}

enum Res {
    Eof,
    Ok,
    ExpectedExecutableDef,
}

fn executable_def(parser: &mut Parser) -> Res {
    parser.skip_ws();
    match parser.current_pair() {
        None => {
            // Return some EOF indicator
            return Res::Eof;
        }
        Some((Token::Name, "query")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(QUERY_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "mutation")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(MUTATION_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "subscription")) => {
            parser.builder.start_node(OPERATION_DEF.into());
            parser.bump_as(SUBSCRIPTION_KEYWORD);
            operation(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::OpenCurly, _)) => {
            parser.builder.start_node(OPERATION_DEF.into());
            selection_set(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        Some((Token::Name, "fragment")) => {
            parser.builder.start_node(FRAGMENT_DEF.into());
            parser.bump_as(FRAGMENT_KEYWORD);
            fragment(parser);
            parser.builder.finish_node();
            return Res::Ok;
        }
        other => {
            return Res::ExpectedExecutableDef;
        }
    }
}

fn operation(parser: &mut Parser) {
    parser.skip_ws();
    if let Some(Token::Name) = parser.current() {
        parser.bump();
    }

    parser.skip_ws();
    if let Some(Token::OpenParen) = parser.current() {
        variable_defs(parser);
    }

    // TODO: directives

    parser.skip_ws();
    if let Some(Token::OpenCurly) = parser.current() {
        selection_set(parser);
    } else {
        todo!()
        // TODO: Error
    }
}

fn variable_defs(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::OpenParen));
    parser.builder.start_node(VARIABLE_DEFS.into());
    parser.bump();
    loop {
        parser.skip_ws();
        match parser.current() {
            None => break,
            Some(Token::Dollar) => {
                parser.builder.start_node(VARIABLE_DEF.into());
                variable(parser);
                parser.skip_ws();
                match parser.current() {
                    Some(Token::Colon) => {
                        parser.bump();
                    }
                    _ => parser.error("expected :"),
                }
                type_(parser);
                parser.skip_ws();
                if let Some(Token::Equals) = parser.current() {
                    parser.builder.start_node(DEFAULT_VALUE.into());
                    parser.bump();
                    constant_value(parser);
                    parser.builder.finish_node();
                }
                parser.builder.finish_node();
                // todo: default values
            }
            Some(Token::CloseParen) => {
                parser.bump();
                break;
            }
            _ => {
                todo!()
                // TODO: error
            }
        }
    }
    parser.builder.finish_node();
}

fn variable(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Dollar));

    parser.builder.start_node(VARIABLE.into());
    parser.bump();
    match parser.current() {
        Some(Token::Name) => parser.bump(),
        _ => parser.error("expected name"),
    }
    parser.builder.finish_node();
}

fn type_(parser: &mut Parser) {
    parser.skip_ws();
    match parser.current() {
        Some(Token::Name) => {
            parser.builder.start_node(TYPE.into());
            named_type(parser);
            if let Some(Token::Bang) = parser.current() {
                parser.bump();
            }
            parser.builder.finish_node();
        }
        Some(Token::OpenSquare) => {
            parser.builder.start_node(TYPE.into());
            parser.bump();
            type_(parser);
            match parser.current() {
                Some(Token::CloseSquare) => {
                    parser.bump();
                }
                _ => {
                    parser.error("expected ]");
                }
            }
            parser.skip_ws();
            if let Some(Token::Bang) = parser.current() {
                parser.bump()
            }
            parser.builder.finish_node();
        }
        _ => {
            parser.error("expected name or [");
        }
    }
    parser.skip_ws();
}

fn named_type(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(NAMED_TYPE.into());
    parser.bump();
    parser.builder.finish_node();
}

fn selection_set(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::OpenCurly));

    parser.builder.start_node(SELECTION_SET.into());
    parser.bump();

    loop {
        parser.skip_ws();
        match parser.current() {
            None => {
                parser.error("expected selection");
                break;
            }
            Some(Token::Name) => {
                parser.builder.start_node(SELECTION.into());
                field_selection(parser);
                parser.builder.finish_node();
            }
            Some(Token::CloseCurly) => {
                parser.bump();
                break;
            }
            Some(Token::Spread) => {
                parser.builder.start_node(SELECTION.into());
                match parser.peek_next_str_non_ws() {
                    Some("on") => inline_fragment(parser),
                    _ => fragment_spread(parser),
                }
                parser.builder.finish_node();
            }
            _ => {
                // TODO: is this good? not sure it is..
                parser.error("expected selection");
                break;
            }
        }
    }

    parser.builder.finish_node();
}

fn field_selection(parser: &mut Parser) {
    assert_eq!(parser.current(), Some(Token::Name));
    parser.builder.start_node(FIELD_SELECTION.into());
    if let Some(Token::Colon) = parser.peek_next_non_ws() {
        parser.builder.start_node(ALIAS.into());

        // Take the name & colon
        parser.bump();
        parser.skip_ws();
        parser.bump();
        parser.builder.finish_node();
    }

    parser.skip_ws();
    match parser.current() {
        Some(Token::Name) => {
            parser.bump();
        }
        _ => parser.error("expected name"),
    }

    parser.skip_ws();
    if let Some(Token::OpenParen) = parser.current() {
        arguments(parser);
    }

    parser.skip_ws();
    if let Some(Token::At) = parser.current() {
        todo!()
        // TODO: parse directives
    }

    parser.skip_ws();
    if let Some(Token::OpenCurly) = parser.current() {
        selection_set(parser)
    }

    parser.builder.finish_node();
}

#[cfg(test)]
mod tests {
    use std::{fs::File, io::Read};

    use rstest::rstest;

    use super::*;

    #[test]
    fn test_simple_parse() {
        let query = r#"
query {
    posts {
        title
        content
    }
}
"#;
        let result = parse(query);
        assert_eq!(result.errors, vec![]);
        assert_eq!(result.green_node.to_string(), query);
        insta::assert_debug_snapshot!(result.syntax());
    }

    #[rstest]
    #[case::fragment("tests/queries/fragment.graphql")]
    #[case::fragment_spread("tests/queries/fragment_spread.graphql")]
    #[case::inline_fragment("tests/queries/inline_fragment.graphql")]
    #[case::minimal("tests/queries/minimal.graphql")]
    #[case::minimal_mutation("tests/queries/minimal_mutation.graphql")]
    #[case::minimal_query("tests/queries/minimal_query.graphql")]
    #[case::named_query("tests/queries/named_query.graphql")]
    #[case::nested_selection("tests/queries/nested_selection.graphql")]
    #[case::query_aliases("tests/queries/query_aliases.graphql")]
    #[case::query_vars("tests/queries/query_vars.graphql")]
    #[case::query_arguments("tests/queries/query_arguments.graphql")]
    #[case::query_float_arguments("tests/queries/query_float_arguments.graphql")]
    #[case::query_list_argument("tests/queries/query_list_argument.graphql")]
    #[case::query_nameless_vars("tests/queries/query_nameless_vars.graphql")]
    // Commenting this one out because it requires comma ignores & comments
    //#[case::query_vars("tests/queries/query_nameless_vars_multiple_fields.graphql")]
    #[case::query_nameless_vars_multiple_fields_canonical(
        "tests/queries/query_nameless_vars_multiple_fields_canonical.graphql"
    )]
    #[case::query_object_argument("tests/queries/query_object_argument.graphql")]
    #[case::string_literal("tests/queries/string_literal.graphql")]
    #[case::triple_quoted_literal("tests/queries/triple_quoted_literal.graphql")]
    #[case::query_var_default_float("tests/queries/query_var_default_float.graphql")]
    #[case::query_var_default_list("tests/queries/query_var_default_list.graphql")]
    #[case::query_var_default_object("tests/queries/query_var_default_object.graphql")]
    #[case::query_var_default_string("tests/queries/query_var_default_string.graphql")]
    #[case::query_var_defaults("tests/queries/query_var_defaults.graphql")]
    fn test_query_file(#[case] file: String) {
        let mut query = String::new();
        File::open(file)
            .unwrap()
            .read_to_string(&mut query)
            .unwrap();

        let result = parse(&query);

        assert_eq!(result.errors, vec![]);
        assert_eq!(result.green_node.to_string(), query);
        // TODO: Determine the snapshot name from the file - to avoid ordering issues...
        insta::assert_debug_snapshot!(result.syntax());
    }
}
