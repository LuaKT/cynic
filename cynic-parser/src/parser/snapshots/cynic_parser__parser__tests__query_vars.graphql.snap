---
source: cynic-parser/src/parser/mod.rs
expression: result.syntax()

---
DOCUMENT@0..38
  OPERATION_DEF@0..37
    OPERATION_TYPE@0..5
      QUERY_KEYWORD@0..5 "query"
    WHITESPACE@5..6 " "
    NAME@6..9 "Foo"
    VARIABLE_DEFS@9..25
      OPEN_PAREN@9..10 "("
      VARIABLE_DEF@10..24
        VARIABLE@10..14
          DOLLAR@10..11 "$"
          NAME@11..14 "arg"
        COLON@14..15 ":"
        WHITESPACE@15..16 " "
        TYPE@16..24
          NAMED_TYPE@16..24
            NAME@16..24 "SomeType"
      CLOSE_PAREN@24..25 ")"
    WHITESPACE@25..26 " "
    SELECTION_SET@26..37
      OPEN_CURLY@26..27 "{"
      LINE_TERMINATOR@27..28 "\n"
      WHITESPACE@28..30 "  "
      FIELD_SELECTION@30..36
        NAME@30..35 "field"
        LINE_TERMINATOR@35..36 "\n"
      CLOSE_CURLY@36..37 "}"
  LINE_TERMINATOR@37..38 "\n"

